### PR TITLE
Stable: extract rate provider functionality

### PR DESCRIPTION
**Pull request by EndymionJkb**
Extract the last semantically distinct non-immutable storage and handling into a separate contract: had to introduce parameter munging functions and pass structs too avoid stack issues.

There is still a fair amount of rate provider processing in the main contract, but this is because these functions also call into StableStorage, and I didn't want the subcontracts to inherit from each other. (I suppose this could be factored out further using interfaces if we really wanted a complete separation.)

### CODE REVIEW

File: pkg/pool-stable-phantom/contracts/StablePhantomPool.sol [hunk 1]
```diff
@@ -1172,6 +1122,29 @@ contract StablePhantomPool is
...
+        returns (StablePoolStorage.StorageParams memory)
+    {
+        return
+            StablePoolStorage.StorageParams(
+                _insertSorted(params.tokens, IERC20(this)),
+                params.rateProviders,
+                params.exemptFromYieldProtocolFeeFlags
+            );
+    }
+
```
Comment (nventuro) [inline pkg/pool-stable-phantom/contracts/StablePhantomPool.sol:1147]: I'd place this right below the constructor

File: pkg/pool-stable-phantom/contracts/StablePoolRates.sol [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-interfaces/contracts/pool-utils/IRateProvider.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/InputHelpers.sol";
+import "@balancer-labs/v2-pool-utils/contracts/rates/PriceRateCache.sol";
+
+abstract contract StablePoolRates {
```
Comment (nventuro) [inline pkg/pool-stable-phantom/contracts/StablePoolRates.sol:23]: I think it's fine to have this inherit from storage to be honest. That'll let us add many other functions, such as `getTokenRateCache`, `getTokenRate`, `_getAdjustedBalances`, `setTokenRateCacheDuration`, etc. As is, this module is not too useful for testing responsability removal purposes.

Down the line, if say we had multiple stable pools, we could then abstract away storage functions into virtual ones (like we do in BaseWeightedPool), but we don't need to worry about that now.

File: pkg/pool-stable-phantom/contracts/StablePhantomPool.sol (hunk 1 above)
Comment (EndymionJkb) [reply pkg/pool-stable-phantom/contracts/StablePhantomPool.sol:1147]: I had it there, then moved it to helpers

File: pkg/pool-stable-phantom/contracts/StablePoolRates.sol (hunk 2 above)
Comment (EndymionJkb) [reply pkg/pool-stable-phantom/contracts/StablePoolRates.sol:23]: Ok; can try it that way. (Getting very close to the original "aggressive" PR, though the helper trick is an improvement.)

**Review by nventuro — COMMENTED** [review]
Looking good! What about the `_getAdjustedBalances` family of functions, and potentially `_scalingFactors`?

File: pkg/pool-stable-phantom/contracts/StablePoolRates.sol [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    // [ uint32  |  uint32  |  uint96  |   uint96     ]
+
+    mapping(IERC20 => bytes32) internal _tokenRateCaches;
+
+    event TokenRateCacheUpdated(IERC20 indexed token, uint256 rate);
+    event TokenRateProviderSet(IERC20 indexed token, IRateProvider indexed provider, uint256 cacheDuration);
+
+    constructor(StablePoolStorage.StorageParams memory storageParams, RatesParams memory rateParams)
+        StablePoolStorage(storageParams)
+    {
```
Comment (nventuro) [inline pkg/pool-stable-phantom/contracts/StablePoolRates.sol:53]: Let's not call `StablePoolStorage`'s constructor here, and instead leave it up to the parent contract (like we do in e.g. `EOASignaturesValidator`, which doesn't initialize `EIP712`).

File: pkg/pool-stable-phantom/contracts/StablePoolRates.sol [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        returns (
+            uint256 rate,
+            uint256 duration,
+            uint256 expires
+        )
+    {
+        _require(_getRateProvider(token) != IRateProvider(0), Errors.TOKEN_DOES_NOT_HAVE_RATE_PROVIDER);
+
+        rate = _tokenRateCaches[token].getCurrentRate();
+        (duration, expires) = _tokenRateCaches[token].getTimestamps();
```
Comment (nventuro) [inline pkg/pool-stable-phantom/contracts/StablePoolRates.sol:120]: We could expand this to also return the old rate.

Comment (nventuro) [reply pkg/pool-stable-phantom/contracts/StablePoolRates.sol:120]: (this is an effort separate from this PR, but I just noticed this)

**Review by nventuro — COMMENTED** [review]
Looking great! StablePool looks very reasonable after this change.

We should be able to extensively test this new rates module (in a separate PR) without having to make real Pool operatrions. Thanks a lot!

**Pull request merged**